src/charm.py

```diff
@@ -0,0 +1,161 @@
+            return
+        self.log = logging.getLogger(__name__)
+        self.image = OCIImageResource(self, "oci-image")
+
+        try:
+            self.interfaces = get_interfaces(self)
+        except NoVersionsListed as err:
+            self.model.unit.status = WaitingStatus(str(err))
+            return
+        except NoCompatibleVersions as err:
+            self.model.unit.status = BlockedStatus(str(err))
+            return
+
+        for event in [
+            self.on.install,
+            self.on.leader_elected,
+            self.on.upgrade_charm,
+            self.on.config_changed,
+            self.on["kubeflow-profiles"].relation_changed,
+        ]:
+            self.framework.observe(event, self.main)
+
+        self.framework.observe(self.on["ingress"].relation_changed, self.configure_mesh)
+
+    def configure_mesh(self, event):
+        if self.interfaces["ingress"]:
+            self.interfaces["ingress"].send_data(
+                {
+                    "prefix": "/",
+                    "rewrite": "/",
+                    "service": self.model.app.name,
+                    "port": self.model.config["port"],
+                }
+            )
+
+    def main(self, event):
+        try:
+            image_details = self.image.fetch()
+        except OCIImageResourceError as e:
+            self.model.unit.status = e.status
+            self.log.info(e)
+            return
+
+        model = self.model.name
+
+        if not (
+            (kf_profiles := self.interfaces["kubeflow-profiles"])
+            and kf_profiles.get_data()
+        ):
+            self.model.unit.status = WaitingStatus(
+                "Waiting for kubeflow-profiles relation data"
+            )
+            return
+
+        kf_profiles = list(kf_profiles.get_data().values())[0]
+        profiles_service = kf_profiles["service-name"]
+
+        port = self.model.config["port"]
+        profile = self.model.config["profile"]
+
+        self.model.unit.status = MaintenanceStatus("Setting pod spec")
+
+        self.model.pod.set_spec(
+            {
+                "version": 2,
+                "serviceAccount": {
+                    "global": True,
+                    "rules": [
+                        {
+                            "apiGroups": [""],
+                            "resources": ["events", "namespaces", "nodes"],
+                            "verbs": ["get", "list", "watch"],
+                        },
+                        {
+                            "apiGroups": ["", "app.k8s.io"],
+                            "resources": [
+                                "applications",
+                                "pods",
+                                "pods/exec",
+                                "pods/log",
+                            ],
+                            "verbs": ["get", "list", "watch"],
+                        },
+                        {"apiGroups": [""], "resources": ["secrets"], "verbs": ["get"]},
```

Comment:
This probably needs to get updated to match the permissions from these two roles:
https://github.com/kubeflow/manifests/blob/master/apps/centraldashboard/upstream/base/clusterrole.yaml
https://github.com/kubeflow/manifests/blob/master/apps/centraldashboard/upstream/base/role.yaml

We just combine the two of them since there isn't a way of writing a charm that uses both a Role and ClusterRole simultaneously. You can add additional service accounts down in the `kubernetesResources` section, but the workload pod won't use it.